### PR TITLE
User defined Diagrams

### PR DESCRIPTION
**Pull request by roleyfoley**
Adds support for user defined diagrams which are filtered to include specific components based on different rules 

A user configures a diagram with a diagram type which has been implemented by a plugin. Along with this they add Rules to the diagram which specify what is included based on occurrence attributes. 

The initial implementation has 3 rule types 
- Links - Which uses the standard link structure to explicitly specify occurrences you want to include 
- ComponentTypes - includes occurrences which map types defined in the rule 
- ResourceType - Similar to ComponentTypes filter but applied based on the types of resources in occurrences

Once the diagram has been defined the user can generate the diagram using the diagram entrance and specifying the deployment unit as the Id of the diagram 

Breaking Change: This replaces the existing functionality and requires users to define their own diagrams

### CODE REVIEW

File: diagrams/components/lb/setup.ftl [hunk 1]
```diff
@@ -1,17 +1,20 @@
...
-            [@setupIPAddressGroupsRelationhsips
-                entityId=subOccurrence.Core.TypedId
-                IPAddressGroups=solution.IPAddressGroups
-                includeLocal=false
-                inbound=true
-            /]
+        [#if isOccurrenceInDiagram(subOccurrence)]
+            [#local solution = subOccurrence.Configuration.Solution ]
+            [#if (solution.IPAddressGroups![])?has_content ]
+                [@setupIPAddressGroupsRelationhsips
```
Comment (RossMurr4y) [inline diagrams/components/lb/setup.ftl]: This macro has a typo, but a search of the definition shows that it was defined that way. Should we correct that now?

File: diagrams/references/Diagram/id.ftl [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        },
+        {
+            "Names" : "Type",
+            "Description" : "The type of diagram to display the data in",
+            "Types" : STRING_TYPE,
+            "Mandatory" : true
+        },
+        {
+            "Names" : "Rules",
+            "Subobjects" : true,
```
Comment (RossMurr4y) [inline diagrams/references/Diagram/id.ftl]:
```suggestion
            "SubObjects" : true,
```

File: diagrams/references/Diagram/id.ftl [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                    "Types" : STRING_TYPE,
+                    "Values" : [ "Links", "ComponentType", "ResourceType" ],
+                    "Mandatory" : true
+                },
+                {
+                    "Names" : "policy:Links",
+                    "Children" : [
+                        {
+                            "Names" : "Links",
+                            "Subobjects" : true,
```
Comment (RossMurr4y) [inline diagrams/references/Diagram/id.ftl]:
```suggestion
                            "SubObjects" : true,
```

**Review by RossMurr4y — DISMISSED** [review]

File: diagrams/components/lb/setup.ftl (hunk 1 above)
Comment (roleyfoley) [reply diagrams/components/lb/setup.ftl]: Fixed

**Pull request merged**